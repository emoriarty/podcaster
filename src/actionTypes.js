export const FETCH_COUNTRIES = 'itunes/FETCH_COUNTRIES';
export const FETCH_COUNTRIES_SUCCESS = 'itunes/FETCH_COUNTRIES_SUCCESS';
export const FETCH_COUNTRIES_FAIL = 'itunes/FETCH_COUNTRIES_FAIL';

export const FETCH_MEDIA_TYPES = 'itunes/FETCH_MEDIA_TYPES';
export const FETCH_MEDIA_TYPES_SUCCESS = 'itunes/FETCH_MEDIA_TYPES_SUCCESS';
export const FETCH_MEDIA_TYPES_FAIL = 'itunes/FETCH_MEDIA_TYPES_FAIL';

export const FETCH_COMMON_TRANSLATIONS = 'itunes/FETCH_COMMON_TRANSLATIONS';
export const FETCH_COMMON_TRANSLATIONS_SUCCESS = 'itunes/FETCH_COMMON_TRANSLATIONS_SUCCESS';
export const FETCH_COMMON_TRANSLATIONS_FAIL = 'itunes/FETCH_COMMON_TRANSLATIONS_FAIL';

export const FETCH_MEDIA_TYPES_TRANSLATIONS = 'itunes/FETCH_MEDIA_TYPES_TRANSLATIONS';
export const FETCH_MEDIA_TYPES_TRANSLATIONS_SUCCESS = 'itunes/FETCH_MEDIA_TYPES_TRANSLATIONS_SUCCESS';
export const FETCH_MEDIA_TYPES_TRANSLATIONS_FAIL = 'itunes/FETCH_MEDIA_TYPES_TRANSLATIONS_FAIL';
