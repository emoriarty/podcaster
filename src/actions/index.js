export { default as fetchCountries } from './itunes/fetchCountries';
export { default as fetchMediaTypes } from './itunes/fetchMediaTypes';
export { default as fetchCommonTranslations } from './itunes/fetchCommonTranslations';
export { default as fetchMediaTypesTranslations } from './itunes/fetchMediaTypesTranslations';
