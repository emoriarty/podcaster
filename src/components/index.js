export * from './Expansion'
export * from './Layouts'
export * from './ListItems'
export * from './Texts'
