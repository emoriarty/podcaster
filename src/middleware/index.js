export { default as yqlMiddleware } from './yql'
export { default as rssMiddleware } from './rss'
