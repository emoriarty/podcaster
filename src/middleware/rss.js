import { normalize } from 'normalizr'

export const RSS_CALL= '@@rss/CALL'

const rss = store => next => async action => {
  const rssAction = action[RSS_CALL]
  if (typeof rssAction === 'undefined') {
    return next(action)
  }

  const {
    url,
    country,
    params,
    schema,
    types
  } = rssAction

  next({
    type: types.request,
    payload: {
      url,
      params
    }
  })

  let data
  try {
    data = await fetch(`${url}/json`)
      .then(response => response.json())
      .catch(err => { throw err })
  } catch (error) {
    next({
      type: types.fail,
      error
    })
    throw error
  }

  const result = next({
    type: types.success,
    payload: schema ? normalize(data, schema) : response
  })
  return result
}

export default rss
