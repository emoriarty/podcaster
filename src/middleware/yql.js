import yqlJsonp from '../utils/yqlJsonp';
import config from '../../config/app.js';
import { normalize } from 'normalizr';

export const YQL_CALL = '@@yql/CALL';

const yql = store => next => async action => {
  const yqlAction = action[YQL_CALL];
  if (typeof yqlAction === 'undefined') {
    return next(action);
  }
  const {
    resource,
    schema,
    types
  } = yqlAction;

  next({
    type: types.request,
    payload: {
      resource
    }
  });

  let response;
  try {
    response = await yqlJsonp(`${config.itunes.provider.rss}/${resource}.json`);
  } catch (error) {
    next({
      type: types.fail,
      error
    });
    throw error;
  }

  const result = next({
    type: types.success,
    payload: schema ? normalize(response, schema) : response
  });
  return result;
};

export default yql;
