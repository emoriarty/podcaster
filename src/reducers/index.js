import { combineReducers } from 'redux'
import countries from './countries'
import entities from './entities'
import mediaTypes from './mediaTypes'
import settings from './settings'
import translations from './translations'
import podcasts from './podcasts'

const app = combineReducers({
  entities,
  countries,
  mediaTypes,
  settings,
  translations,
  podcasts
})

export default app
