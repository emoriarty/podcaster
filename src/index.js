import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import appReducer from './reducers';
import { fetchCountries } from './actions';

const loggerMiddleware = createLogger();

window.addEventListener('load', () => {
  // jsonp(
  //   'https://itunes.apple.com/search?term=carne&media=podcast',
  //   (err, response) => console.log(err, response)
  // );

  let store = createStore(
    appReducer,
    applyMiddleware(
      thunkMiddleware,
      loggerMiddleware
    )
  );

  console.log('initialState', store.getState());

  const unsubscribe = store.subscribe(() => console.log(store.getState()));

  store
    .dispatch(fetchCountries())
    .then(data => {
      console.log('success', data);
    })
    .catch(reason => {
      console.error(reason);
    });

  unsubscribe();
});
