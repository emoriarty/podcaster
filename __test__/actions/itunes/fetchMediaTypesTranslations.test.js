import { fetchMediaTypesTranslations } from '../../../src/actions'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { yqlMiddleware } from '../../../src/middleware'

jest.mock('../../../src/utils/yqlJsonp')

const mockStore = configureStore([ thunk, yqlMiddleware ])

describe('fetchMediaTypesTranslations', () => {
  it('dispatchs actions properly', () => {
    const store = mockStore({ countries: [] })
    const language = 'en'

    return store.dispatch(fetchMediaTypesTranslations(language))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
      })
  })
})
