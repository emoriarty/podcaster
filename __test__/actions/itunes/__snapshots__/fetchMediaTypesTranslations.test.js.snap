// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchMediaTypesTranslations dispatchs actions properly 1`] = `
Array [
  Object {
    "payload": Object {
      "resource": "data/lang/en/media-types",
    },
    "type": "itunes/FETCH_MEDIA_TYPES_TRANSLATIONS",
  },
  Object {
    "payload": Array [
      Object {
        "canBeExplicit": true,
        "feed_types": Array [
          Object {
            "translation_key": "topaudiobooks",
            "urlPrefix": "https://itunes.apple.com/<%= country_code %>/rss/topaudiobooks/<%= parameters %>",
            "urlSuffix": "xml",
          },
        ],
        "id": "50000024",
        "store": "music",
        "subgenres": Array [
          Object {
            "id": "",
            "translation_key": "all",
          },
          Object {
            "id": "50000041",
            "translation_key": "arts&entertainment",
          },
        ],
        "translation_key": "audiobooks",
      },
    ],
    "type": "itunes/FETCH_MEDIA_TYPES_TRANSLATIONS_SUCCESS",
  },
]
`;
